[do not merge] Add enforce active org guide (#1331)
clerk/clerk-docs: docs/guides/enforce-active-organization.mdx

---
title: Enforce an active organization for users
description: Learn how to use an onboarding like flow to enforce that all users have an active organization.
---

# Enforce an active organization for users

> Clerk's organization feature enables the users of an application to be members of more than one organization at a time. Each membership can have its own role, allowing a user to habe different privileges for each organization.

Reviewer: ```suggestion
Clerk's organization feature enables the users of an application to be members of more than one organization at a time. Each membership can have its own role, allowing a user to have different privileges for each organization.
```

## How do I know if a user is active in an organization?

You can check the `orgId` value on the [Auth](docs/references/nextjs/auth-object#auth-object). If this value is `undefined` then the user is not active in an organization and is instead in their Personal Workspace.

If you check the returned Auth object from [auth()](/docs/references/nextjs/auth) and the user is not active in an organization, it would like this:

```json
{
  "sessionId": "sess_2GaMqUCB3Sc1WNAkWuNzsnYVVEy",
  "userId": "user_2F2u1wtUyUlxKgFkKqtJNtpJJWj",
  "claims": {
    "azp": "http://localhost:3000",
    "exp": 1666622607,
    "iat": 1666622547,
    "iss": "https://clerk.quiet.muskox-85.lcl.dev",
    "nbf": 1666622537,
    "sid": "sess_2GaMqUCB3Sc1WNAkWuNzsnYVVEy",
    "sub": "user_2F2u1wtUyUlxKgFkKqtJNtpJJWj"
  }
}
```

## Enforcing an active organization

<Steps>

>   ### Configuring Middleware

Reviewer 2: ```suggestion
  ### Configure Middleware to protect pages
```

>   Let's start with a Middleware and that many applications could use. This example will do the following:
> 
>   1. Check if the user accessing a private route is signed in. If the user is not, they are redirected to the `/sign-in` page.
>   1. Check if the user accessing a private route is signed in and if the user is then allow them to access that route.
>   1. Any use who is not accessing a public route will get access to the prublic.
> 
>   ```typescript {{ filename: '/middleware.ts' }}
>   import { clerkMiddleware, createRouteMatcher } from '@clerk/nextjs/server'
>   import { NextRequest, NextResponse } from 'next/server'
> 
>   const isPublicRoute = createRouteMatcher(['/', '/sign-in', '/sign-up'])
> 
>   export default clerkMiddleware((auth, req: NextRequest) => {
>     // If the user isn't signed in and the route is private, redirect to sign-in
>     if (!auth().userId && !isPublicRoute(req))
>       return auth().redirectToSignIn({ returnBackUrl: req.url })
> 
>     // If the user is logged in and the route is protected, let them view.
>     if (auth().userId && !isPublicRoute(req)) return NextResponse.next()
>   })
> 
>   export const config = {
>     matcher: ['/((?!.*\\..*|_next).*)', '/', '/(api|trpc)(.*)'],
>   }
>   ```

Reviewer 2: We don't need two separate examples, when the second example is the same code as the first with the addition of the orgId check. We can simply say something like:

```suggestion
The `orgId` value on the `Auth` object can be used to check if a user has an active organization. If this value is `undefined`, then the user is not active in an organization and is in their personal workspace.

To protect your application from users that do not have an active organization, configure your Middleware to check the user's `orgId`. 

The following example configures `clerkMiddleware()` to protect all routes except `/`, `/sign-in`, and `/sign-up`. It will handle both authentication _and_ authorization. If a user is not signed in, they will be redirected to the sign-in page. If a user does not have an active organization, they will be redirected to`/organization-selection`, which you will create in the next step.
```

  The following addition will determine if a user have an active organization or not. This check doesn't determine if the user is accessing a public route or a private route -- it is just checkingi if the user is signed in and is missing an `orgId`. If the user does not have an active organization, then redirect to `/organization-selection`.

  ```typescript {{ filename: '/middleware.ts', mark: [[7, 13]] }}
  import { clerkMiddleware, createRouteMatcher } from '@clerk/nextjs/server'
  import { NextRequest, NextResponse } from 'next/server'

  const isPublicRoute = createRouteMatcher(['/', '/sign-in', '/sign-up'])

  export default clerkMiddleware((auth, req: NextRequest) => {
    // Check if the user is signed in and does not have an auth().orgId

>     // If true, then redirect them to the organization selection route

Reviewer 2: ```suggestion
    // If they don't have an active org, redirect them to `/organization-selection`
```

>     // Include their current route as a redirect_url, which will be handled on the route

Reviewer 2: I don't think this is added yet
```suggestion
```

    if (auth().userId && !auth().orgId && req.nextUrl.pathname !== '/organization-selection') {
      const orgListUrl = new URL('/organization-selection', req.url)
      return NextResponse.redirect(orgListUrl)
    }

>     // If the user isn't signed in and the route is private, redirect to sign-in

Reviewer 2: ```suggestion
    // If the user isn't signed in and the route is protected, redirect them to sign-in
```

    if (!auth().userId && !isPublicRoute(req))
      return auth().redirectToSignIn({ returnBackUrl: req.url })

>     // If the user is logged in and the route is protected, let them view.

Reviewer 2: ```suggestion
    // If the user is signed in and the route is protected, let them view
```

    if (auth().userId && !isPublicRoute(req)) return NextResponse.next()
  })

>   export const config = {
>     matcher: ['/((?!.*\\..*|_next).*)', '/', '/(api|trpc)(.*)'],
>   }

Reviewer 2: we updated the matcher
```suggestion
  export const config = {
  matcher: [
    // Skip Next.js internals and all static files, unless found in search params
    '/((?!_next|[^?]*\\.(?:html?|css|js(?!on)|jpe?g|webp|png|gif|svg|ttf|woff2?|ico|csv|docx?|xlsx?|zip|webmanifest)).*)',
    // Always run for API routes
    '/(api|trpc)(.*)',
  ],
};
```

  ```

  ### Setup the `/organization-selection` route

>   The `/ogranization-selection` route will render the `<OrganizationList />` component. The `hidePeronsal` prop is passed, which will hide the option for a user to select their Personal Worksapce. This will mean that all users, even new users, will need to either create or join and organization.
> 
>   The component will also check the `orgId` from `auth()`. This will be falsy when the user is first redirected to the route. When the user creates or joins an organization, the auth object will be refreshed and `orgId` will have a valid organization id. When that happens the user will be redirected to `/dashboard`.

Reviewer 2: ```suggestion
  Create the `/organization-selection` route, which is where users who do not have an active organization will get redirected to. 
  
  The following example:
  - Uses the `<OrganizationList />` component to list the user's organization memberships. The `hidePeronsal` prop is passed, which will hide the option for a user to select their personal workspace. This will mean that all users, even new users, will need to either create or join an organization.
- Checks the `orgId` from `auth()`. Initially, this will be falsy when the user is redirected to the route. Once the user creates or joins an organization, the `Auth` object is refreshed and `orgId` contains a valid organization ID. At this point, the user will be redirected to `/dashboard`.
```

  ```tsx {{ filename: '/app/organization-selection/page.tsx' }}
  import { OrganizationList } from '@clerk/nextjs'
  import { auth } from '@clerk/nextjs/server'
  import { redirect } from 'next/navigation'

  export default async function OrganizationSelectionPage() {
    const { orgId } = auth()

>     // Once the `orgId` is set, redirect the user. In example the user will be redirected to `/dashboard`

Reviewer 2: ```suggestion
    // Once the user has an active org, redirect them
```

    if (orgId) {
      redirect('/dashboard')
    }

    return (
      <>
        <h1>Select an organization</h1>
        <OrganizationList hidePersonal />
      </>
    )
  }
  ```

  ### Improving the user experience with redirects

>   When a user does not have an active organization and is redirected, we can add a `redirect_url` to the URL and then later handle the direct on the `/organization-selection` route.

Reviewer 2: ```suggestion
  To provide a more seamless user experience, you can add a `redirect_url` to the URL when redirecting to the `/organization-selection` route. This will allow the user to be redirected back to the route they were trying to access, once they have created or joined an organization.
```

>   ### Updating Middleware
> 
>   Let's start with the changes to Middleware. The Middleware will now read the `req.nextUrl.href` and URL endcode it, and then add that as a `redirect_url` search parameter.

Reviewer 2: ```suggestion
  #### Update the Middleware

  Modify `clerkMiddleware()` to include the `redirect_url` search parameter in the URL when redirecting to the `/organization-selection` route. This will allow that route to consume the search parameter and use it to redirect the user back to the route they were trying to access once they have created or joined an organization.
```

  ```typescript {{ filename: '/middleware.ts', mark: [[11, 12]] }}
  import { clerkMiddleware, createRouteMatcher } from '@clerk/nextjs/server'
  import { NextRequest, NextResponse } from 'next/server'

  const isPublicRoute = createRouteMatcher(['/', '/sign-in', '/sign-up'])

  export default clerkMiddleware((auth, req: NextRequest) => {
    // Check if the user is signed in and does not have an auth().orgId

>     // If true, then redirect them to the organization selection route

Reviewer 2: ```suggestion
    // If they don't have an active org, redirect them to `/organization-selection`
```

    // Include their current route as a redirect_url, which will be handled on the route
    if (auth().userId && !auth().orgId && req.nextUrl.pathname !== '/organization-selection') {
      const redirectUrl = encodeURIComponent(req.nextUrl.href)
      const orgListUrl = new URL(`/organization-selection?redirect_url=${redirectUrl}`, req.url)
      return NextResponse.redirect(orgListUrl)
    }

>     // If the user isn't signed in and the route is private, redirect to sign-in

Reviewer 2: ```suggestion
    // If the user isn't signed in and the route is protected, redirect them to sign-in
```

    if (!auth().userId && !isPublicRoute(req))
      return auth().redirectToSignIn({ returnBackUrl: req.url })

>     // If the user is logged in and the route is protected, let them view.

Reviewer 2: ```suggestion
    // If the user is signed in and the route is protected, let them view
```

    if (auth().userId && !isPublicRoute(req)) return NextResponse.next()
  })

>   export const config = {
>     matcher: ['/((?!.*\\..*|_next).*)', '/', '/(api|trpc)(.*)'],
>   }

Reviewer 2: ```suggestion
  export const config = {
  matcher: [
    // Skip Next.js internals and all static files, unless found in search params
    '/((?!_next|[^?]*\\.(?:html?|css|js(?!on)|jpe?g|webp|png|gif|svg|ttf|woff2?|ico|csv|docx?|xlsx?|zip|webmanifest)).*)',
    // Always run for API routes
    '/(api|trpc)(.*)',
  ],
};
```

  ```

>   ### Updating `/organization-selection`

Reviewer 2: ```suggestion
  #### Update `/organization-selection`
```

>   With Middleware adding the `redirect_url` search parameter to the URL, the route can then consume that and use it once the user has selected or created an organization.
> 
>   First, modify the component parameters to ready the search parameters. Additionally indicate that the parameter expected is `redirect_url` with a type of string.
> 
>   Second, once `orgId` is truthy check `redirect_url` was passed as a search parameter. If a `redirect_url` was passed then use that value, after URL decoding it, for the redirect. If there is no `redirect_url` then use `/dashboard` as a fallback.

Reviewer 2: Removing repetition, and cleaning up the explanation here.

```suggestion
  Modify the `orgId` check to check if the `redirect_url` search parameter is present. If it is, redirect the user back to that route. If it is not present, redirect the user to a fallback. In this example, the fallback is `/dashboard`.
```

  ```tsx {{ filename: "/app/organization-selection", mark: [[6, 10], [13,21]]}}

  import { OrganizationList } from "@clerk/nextjs";
  import { auth } from "@clerk/nextjs/server";
  import { redirect } from "next/navigation";

  export default async function OrganizationSelectionPage({
    searchParams
  }: {
    searchParams: { redirect_url: string }
  }) {
    const { orgId } = auth()

    if (orgId) {

>       // If the orgId is truthy, then redirect the user to the redirect_url if present

Reviewer 2: ```suggestion
      // If the user has an active org, redirect them to the redirect_url if present
```

      if (searchParams.redirect_url) {
        redirect(decodeURIComponent(searchParams.redirect_url))
      }

      // If there is no redirect_url, redirect to /dashboard
      redirect("/dashboard")
    }

    return (
      <>
        <h1>Select an organization</h1>
        <OrganizationList hidePersonal />
      </>
    )

  }
  ```
</Steps>

## Wrap up

> This will now reliably detect any user without an active organization (`orgId`), whether that user is someone who has just signed up or someone who organization was deleted or who left an organization. Any user without an active organization will be redirected and forced to create or select and organization.

Reviewer 2: ```suggestion
Your application will now reliably detect any user without an active organization (`orgId`), whether that user is someone who has just signed up or someone whose organization was deleted or who left an organization. Any user without an active organization will be redirected and forced to create or join an organization.
```

> The route can modified to fit the needs of your application. You could configure it so the user can join an organization based on [email domain](/docs/organizations/verified-domains), or require a user to subscribe to a plan in your application before creating an organization.

Reviewer 2: ```suggestion
The route can modified to fit the needs of your application. You could configure it so that the user can join an organization based on an [email domain](/docs/organizations/verified-domains), or require a user to subscribe to a plan in your application before creating an organization.
```
